Add second job definition defaulting to hosted RV image

jeancochrane opened the pull request:
## Overview

If the user does not specify a `RepositoryName` to create for use in pushing/pulling RV images, don't create the ECR repo and instead pull the latest RV image from Quay.io.

Closes https://github.com/azavea/raster-vision-aws/issues/12.

## Notes

- This PR branches off of #2.

## Testing instructions

### 1. Verify template syntax

- Confirm that the template syntax is valid:

```
aws cloudformation validate-template --template-body file://cloudformation/template.yml
```

### 2. Create job definition from ECR repo

- Log into the R&D AWS console
- Navigate to `CloudFormation > Create Stack`
- In the `Choose a template field`, select `Upload a template to Amazon S3` and upload the template in `cloudformation/template.yml`
- Where available, use the default parameters. Specify the following required parameters:
    - `Stack Name`: `TestingCloudFormation`
    - `Slug`: A slug of your choosing
    - `VPC`: `vpc-074e8ae6256f7c1ca` (the first one on the dropdown)
    - `Subnets`: `subnet-033d2cbe8268c3067` (the first one on the dropdown)
    - `SSH Key Name`: `raster-vision` (doesn't matter unless you want to shell into the instance)
- Specify a name and image tag for the ECR repo:
    - `Repository Name`: The name for your ECR repo (e.g. `raster-vision-gpu-testing`)
    - `Image Tag`: The tag for a Docker image (e.g. `latest`)
- Accept all default options on the Options screen
- Accept `I acknowledge that AWS CloudFormation might create IAM resources with custom names` on the Review screen
- Create the stack and confirm that it builds correctly
- Navigate to `Batch > Job definitions > ${slug}RasterVision${environment}CustomJobDefinition`
- Under the `Resource requirements` section, confirm that the job definition is pointing to the URI of the new ECR repo

### 3. Modify Stack to use hosted RV image

- Navigate to `CloudFormation > ${StackName} > Actions > Update stack`
- Select `Use current template` and hit `Next`
- Clear out the values from the `Repository Name` and `Image Tag` fields
- Accept all Options and update the stack
- Navigate to `Batch > Job definitions > ${slug}RasterVision${environment}HostedJobDefinition`
- Under the `Resource requirements` section, confirm that the job definition is pointing to the latest RV image on Quay.io

Commit:
Add second job definition defaulting to hosted RV image

If the user does not specify a 'RepositoryName' to create for use
in pushing/pulling RV images, don't create the ECR repo and instead
pull the latest RV image from Quay.io.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -24,11 +24,14 @@ Metadata:
+          default: Docker Image Configuration
         Parameters:
           - RepositoryName
-          - ImageName
```
As you'll see lower down, `ImageName` isn't actually used in the ECR repo URI, so I removed it from the parameter list.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -160,6 +161,10 @@ Parameters:
 
+Conditions:
+  UseHostedDockerImage: !Equals [!Ref RepositoryName, ""]
+  UseCustomDockerImage: !Not [!Equals [!Ref RepositoryName, ""]]
```
We might be able to get away with avoiding this explicit condition and instead using `!Not [UseHostedDockerImage]`, but I figured being a little bit more verbose is useful here anyway.

rbreslow replied:
I would rename this to say `ContainerImage` vs `DockerImage`.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -298,13 +304,37 @@ Resources:
+      Type: Container
+      JobDefinitionName: !Join ['', [!Ref Slug, 'RasterVision', !Ref Environment, 'CustomJobDefinition']]
+      ContainerProperties:
+        Image: !Sub "${AWS::AccountId}.dkr.ecr.${AWS::Region}.amazonaws.com/${RepositoryName}:${ImageTag}"
```
Unfortunately, the `Repository` resource [does not expose an attribute corresponding to the URI of the repo](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-repository.html#aws-resource-ecr-repository-returnvalues), meaning that we have to construct the URI by hand here.

jeancochrane replied:
Note that `ImageName` isn't required for this URI string, since ECR uses the repository name to refer to the image.

rbreslow replied:
Nice.

rbreslow commented on cloudformation/template.yml:
```diff
@@ -24,11 +24,14 @@ Metadata:
       -
         Label:
-          default: Docker Container Configuration
+          default: Docker Image Configuration
```
I would combine this section with the following container configuration block.

rbreslow commented on cloudformation/template.yml:
```diff
@@ -24,11 +24,14 @@ Metadata:
           - ImageTag
+      -
+        Label:
+          default: Docker Container Configuration
```
I would change the wording of this to "Container Image Properties" or "Container Image Configuration."

rbreslow commented on cloudformation/template.yml:
```diff
@@ -139,18 +140,18 @@ Parameters:
+    Description: >
+      (Optional) Specifies the name of an ECR repository to create for use in
+      pushing and pulling images -- if empty, pulls the latest Raster Vision
+      GPU image from Quay.io instead
```
Maybe, "GPU-based Raster Vision container image?"

rbreslow commented on cloudformation/template.yml:
```diff
@@ -300,13 +306,37 @@ Resources:
+        Image: !Sub "${AWS::AccountId}.dkr.ecr.${AWS::Region}.amazonaws.com/${RepositoryName}:${ImageTag}"
+        Vcpus: !Ref InstanceVCPUs
+        Memory: !Ref InstanceMemory
+        Volumes:
```
Here's an example of a Raster Vision GPU job definition (from when we used to use the AWS CLI):

https://github.com/azavea/raster-vision/blob/ff82e5c0122bbe6af55e40c4e69336813d926767/deployment/batch/job_def_gpu.json

I would double check everything here lines up. You can also look at the batch compute environment and queue configurations at this point in history of the repository. I think seeing it all in JSON is useful.

jeancochrane replied:
Nice, that's a helpful resource.

rbreslow reviewed (approved):
Looks good 👍. 

The string substitution you did in the image URL is pretty nice looking.

rbreslow commented:
Didn't think this fit into the review, but wanted to share anyway. 

I've been trying to be agnostic to Docker when referring to containers, because there really isn't such a thing as a Docker image, and there are many alternatives to the Docker tooling. For example, the Docker tooling builds container images that meet the Open Container Initiative (OCI) [specification](https://github.com/opencontainers/image-spec), and can be ran with any container runtime or container engine.

I'll drop some knowledge from @fatherlinux as well:

> Containers do not run ON docker. Containers are processes -- they run on a container host. Containers are Linux.

>The docker daemon is one of the many user space tools/libraries that talks to the kernel to set up containers.

Commit:
Use "Container" instead of "Docker"

Remove references to Docker when talking about containers. In addition,
adjust the default vCPU count and memory limit for containers in the
input variables.

jeancochrane commented:
That's good context RE: containers -- Docker has strong enough branding that it's easy to get by thinking that they invented the whole architecture.
